apps/snmp/snmp_stream.c

```diff
@@ -0,0 +1,475 @@
+};
+/* Linked list of sockets used to listen for events
+ */
+static struct stream_socket *STREAM_SOCKETS = NULL;
+
+/*! Read smiv2:oid for given YANG node
+ *
+ * @param[in]  ys       YANG node
+ * @param[out] objid    oid
+ * @param[out] objidlen oid length
+ * @retval     0  OK
+ * @retval    -1  Error
+ */
+int
+get_oid_for_yang_node(yang_stmt *ys, oid *objid, size_t *objidlen)
+{
+    int   retval = -1;
+    int   exist = 0;
+    char *oidstr = NULL;
+
+    if (yang_extension_value_opt(ys, "smiv2:oid", &exist, &oidstr) < 0)
+        goto done;
+    if (exist == 0 || oidstr == NULL){
+        clixon_debug(CLIXON_DBG_SNMP, "oid not found as SMIv2 yang extension of %s", yang_argument_get(ys));
+        goto done;
+    }
+    if (snmp_parse_oid(oidstr, objid, objidlen) == NULL){
+        clixon_err(OE_SNMP, errno, "snmp_parse_oid");
+        goto done;
+    }
+
+    retval = 0;
+done:
+    return retval;
+}
+
+/*! Add snmp trap v2 oid to the variable list and bind the given notification oid as value.
+ * 
+ * @param[in]  var_list   Variable list
+ * @param[in]  notify_oid OID of the notification to set as value
+ * @param[in]  oid_len    Length of notify_oid
+ */
+void 
+add_snmp_trapv2_oid(netsnmp_variable_list **var_list, oid *notify_oid, int oid_len)
+{
+    snmp_varlist_add_variable(
+        var_list,
+        notificationOid,
+        OID_LENGTH(notificationOid),
+        ASN_OBJECT_ID,
+        (const unsigned char*)notify_oid, oid_len * sizeof(oid)
+    );
+}
+
+/*! Add snmp var binding for all child nodes of cxparent.
+ * 
+ * All child elements are read recursively. A var binding is created for each LEAF element
+ * with an existing oid and added to the snmp variable list.
+ * 
+ * @param[in]  var_list Variable list
+ * @param[in]  cxparent Parent xml node
+ * @param[in]  ys       YANG node
+ */
+int
+add_snmp_var_bindings(netsnmp_variable_list **var_list, cxobj *cxparent, yang_stmt *ys)
+{
+    int             retval = -1;
+    int             ret;
+    cxobj          *xncont = NULL; /* notification context xml */
+    yang_stmt      *ychild;
+    char           *xmlstr = NULL;
+    char           *body = NULL;
+    int             asn1type;
+    char           *reason = NULL;
+    unsigned char  *snmpval = NULL;
+    size_t          snmplen = 0;
+    oid             objid[MAX_OID_LEN] = {0,}; /* Leaf */
+    size_t          objidlen = MAX_OID_LEN;
+
+    while ((xncont = xml_child_each(cxparent, xncont, CX_ELMNT)) != NULL){
+        char *node_name = xml_name(xncont);
+        if ((ychild = yang_find(ys, 0, node_name)) != NULL) {
+            switch (yang_keyword_get(ychild)){
+                case Y_LEAF: /* only Y_LEAF type is supported here */
+                    if (get_oid_for_yang_node(ychild, objid, &objidlen) < 0)
+                        goto done;
+                    if ((body = xml_body(xncont)) != NULL){
+                        if ((ret = type_xml2snmp_pre(body, ychild, &xmlstr)) < 0)
```

Comment:
Memory leak: xmlstr needs to be free:d before a new loop instance. It is freed at the end of the function, but if the while loop iterates more than once the previous value of xmlstr is overwritten.
(tip: run make memsnmp in docker/test)